### PR TITLE
Add Continue/Break feature to loops and iterators

### PR DESCRIPTION
**Pull request by dzjin**
This work ties together iterators and loops by extending both ideas from base classes - BaseIterator and BaseLoop.

By having central logic, we're now able to add Continue/Break to loops.

General trick - Continue/Break activities trigger a `callContext.fail()` with a `reason`, where `reason` is an instance of a Continue or Break event.

Addendum: Calling Break/Continue while inside a Try will bubble up past Try.

### CODE REVIEW

**Comment by wmurray8989** [comment]
What happens if a break or continue is called outside of a loop?

**Comment by dzjin** [comment]
@wmurray8989 Should just fail the workflow

**Review by mollyhenderson — APPROVED** [review]
barf

File: lib/es6/activities/baseIterator.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    const remainingItems = this._remainingItems;
+    const iterator = this._iterator;
+    if (remainingItems && remainingItems.length) {
+      if (this.parallel) {
+        const bodies = this._bodies;
+        const pack = [];
+        let idx = 0;
+        while (remainingItems.length && idx < bodies.length) {
+          const item = remainingItems[0];
+          remainingItems.splice(0, 1);
```
Comment (rsnickell) [inline lib/es6/activities/baseIterator.js:77]: You could also use the splice() return value and remove the previous statement. Another option is to use remainingItems.shift() to remove the head element.

File: lib/es6/activities/for.js [hunk 2]
```diff
@@ -1,91 +1,74 @@
...
+        if (!_.isNumber(to)) {
+            callContext.fail(new TypeError(`For activity's to value '${to}' is not a number.`));
+            return;
+        }
+        if (!_.isNumber(step)) {
+            callContext.fail(new TypeError(`For activity's step value '${step}' is not a number.`));
+            return;
+        }
+        let current;
+        if (_.isNull(this[varName])) {
```
Comment (rsnickell) [inline lib/es6/activities/for.js:58]: ~Can someone assign the varName to be null? ~

Ah, this is the value you're null checking.

File: lib/es6/common/index.js [hunk 3]
```diff
@@ -4,5 +4,6 @@ module.exports = {
     asyncHelpers: require("./asyncHelpers"),
     SimpleProxy: require("./simpleProxy"),
     is: require("./is"),
-    converters: require("./converters")
+    converters: require("./converters"),
+    events: require('./events')
```
Comment (rsnickell) [inline lib/es6/common/index.js]: order and consistent use of quotes please.

File: lib/es6/activities/baseIterator.js (hunk 1 above)
Comment (dzjin) [reply lib/es6/activities/baseIterator.js:77]: I agree, but I'm going to leave as-is - reason being that this was a refactor/port of W4N `@For` code, and I don't really want to mess with it.

File: lib/es6/activities/baseIterator.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+class BaseIterator extends Activity {
+  constructor() {
+    super();
+    this.items = null;
+    this.varName = 'item';
+    this.parallel = false;
+    this._bodies = null;
+  }
+  initializeStructure() {
+    if (this.parallel) {
```
Comment (greenarrowdb) [inline lib/es6/activities/baseIterator.js:17]: Are we even exposing the parallel activity?

**Review by greenarrowdb — APPROVED** [review]
I question how much we need to deal with parallel. But overall it looks good.

**Review by davidlandreman — COMMENTED** [review]
One thing worth discussing briefly is that workflow-4-node is based on the ideas of MS Workflow Foundation which doesn't have this capability:  https://docs.microsoft.com/en-us/dotnet/framework/windows-workflow-foundation/samples/emulating-breaking-in-a-while-activity 

With that said I think I am fine with this.

**Review by davidlandreman — APPROVED** [review]

**Pull request merged**